nemo/collections/asr/parts/k2/graph_decoders.py

```diff
@@ -0,0 +1,210 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Optional, Tuple, Union
+
+import k2
+import torch
+
+from nemo.collections.asr.parts.k2.topologies import build_topo
+from nemo.collections.asr.parts.k2.utils import (
+    create_supervision,
+    intersect_with_self_loops,
+    invert_permutation,
+    load_graph,
+    make_blank_first,
+    prep_padded_densefsavec,
+    shift_labels_inpl,
+)
+from nemo.utils import logging
+
+
+class DecoderConf:
+    """Graph decoder config.
+    Typically contains the pruned intersection parameters.
+    """
+
+    def __init__(
+        self, search_beam=20.0, output_beam=10.0, min_active_states=30, max_active_states=10000, **kwargs,
+    ):
+        self.search_beam = search_beam
+        self.output_beam = output_beam
+        self.min_active_states = min_active_states
+        self.max_active_states = max_active_states
+
+
+class _BaseDecoder(object):
+    """Base graph decoder.
+    Typically uses the same parameters as for the corresponding loss function.
+    This implementation does not initialize decode_graph, 
+    thus its constructor must not be called explicitly.
+    Any custom decoder should implement this class and describe decode_graph initialization.
+    """
+
+    def __init__(
+        self,
+        num_classes: int,
+        blank: int,
+        intersect_pruned: bool = False,
+        topo_type: str = "default",
+        topo_with_selfloops: bool = True,
+        device: torch.device = torch.device("cpu"),
+        **kwargs,
+    ):
+        self.num_classes = num_classes
```

Comment:
K2 import guard check here.

Reply:
Done in [8d3e482](https://github.com/NVIDIA/NeMo/pull/3351/commits/8d3e48215629824b59a020526a0444c944b2d1c2)